text/3064-bitfield.md

```diff
@@ -0,0 +1,401 @@
+
+    ```rust
+    #[repr(C)]
+    struct X {
+        a: BitField<u32, 6>,
+    }
+    ```
+
+    where `BitField` would be some kind of lang item that ensures that
+
+    - the address of a field with such a type cannot be taken
+    - the type and width parameters are correct.
+
+    This type would then have the following API:
+
+    ```rust
+    impl<T, const N: usize> BitField<T, N> {
+        fn new(t: T) -> Self;
+        fn wrapping_new(t: T) -> Self;
+        fn get(self) -> T;
+        fn set(&mut self, t: T);
+        fn wrapping_add(self, t: T) -> Self;
+        ...
+    }
+
+    impl<T, const N: usize> AddAssign<T> for BitField<T, N> {
+    }
+    ```
+
+    The layout of `BitField<T, N>` would be the layout of `T` when used outside
+    a struct.
+
+    This could then be used like so:
+
+    ```rust
+    fn f(x: X) {
+        x.a.set(1);
+        x.a += 1;
+        x.a = x.a.wrapping_add(12);
+        println!("{}", x.a.get());
+    }
+    ```
+
+    There is some magic happening here:
+
+    - `x.a += 1` cannot be written as `AddAssign::add(&mut x.a, 1)` because the
+      address of `x.a` cannot be taken. The compiler would figure it out when
+      the lhs is a struct field.
+    - When the assignment happens, the compiler writes the bits on the rhs to
+      the correct positions in the struct.
+
+    This is a more structured solution but would need to be fleshed out.
+
+# Prior art
+[prior-art]: #prior-art
+
+- C
+- The author of Zig [proposes][zig-proposes] a different syntax that basically
+  boils down to arbitrarily sized integer types:
+
+  ```rust
+  struct X {
+      x: i9,
+  }
+  ```
+
+  But this proposal is flawed because, in C, both the width and the underlying
+  type influence the layout. The Zig proposal throws the underlying type away.
+
+  ```c
+  struct X {
+      char a[3];
+      int b:9;
+  };
+
+  struct Y {
+      char a[3];
+      long b:9;
+  };
+  ```
+
+  These structs have different layouts on `x86_64-unknown-linux-gnu`.
+
+[zig-proposes]: https://andrewkelley.me/post/a-better-way-to-implement-bit-fields.html
+
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- On Windows, Clang and GCC produce different layouts for packed structs:
+
+    ```c
+    #include <stdio.h>
+    
+    struct __attribute__((packed, ms_struct)) X {
+    	char a;
+    	int b:1;
+    	char c;
+    };
+    
+    int main() {
+    	printf("%ld\n", sizeof(struct X));
+    }
+    ```
+
+    ```
+    ~$ gcc test.c && ./a.out
+    6
+    ~$ clang test.c && ./a.out   
+    12
+    ```
+
+    (`ms_struct` is the default on Windows.)
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+- Adding some way to perform wrapping assignment.
+- Relaxing the requirement that `N` must be a literal.
+- Bit-fields with types that are enums with integer representation.
+- Bit-fields with types that are transparent wrapper structs around valid
+  bit-field types.
```

Comment:
```suggestion
  bit-field types.
- Have a way to specify which compiler the layout should be compatible with (gcc, clang, MSVCC, etc.)
```

Reply:
I'm currently working on a comprehensive RFC that will bring various enhancements to repr(C) types including allowing you to opt into MSVC layout on MinGW targets. But that RFC will still take a few more weeks until it's ready.